### ISSUE TITLE
Cross Encoders returning low scores on fairly similar texts

### ISSUE THREAD

**Issue by krumeto** [UKPLab/sentence-transformers#2874]
I noticed that cross encoders return fairly low scores on arguably, similar short texts. One of the tested models returned low scores on the same text as well. I wonder if that behaviour is well-documented? 

To me, at least, it was surprising to see the scores for similar short texts and made me rethink my priors a bit. The scores in the typical retrieval scenario are as expected, but I was under the impression that cross encoders are the gold standard for text similarity and it appears that is not universally the case.

It was also surprising to me that arguably the most logical results are those of the, fairly old and small, `cross-encoder/ms-marco-MiniLM-L-6-v2`. What's additionally surprising is that I'd expect such behaviour to be the result of training sets tailored for retrieval, but then why is precisely that model behaving logically (when it was trained on THE retrieval dataset:)?

If it is a fairly known phenomenon, please feel free to close the issue outright. Thank you for any observations in advance.

```{python}
from sentence_transformers import CrossEncoder
from torch import nn
import pandas as pd

same = ['Freddie was a great singer', "Freddie was a great singer"]
similar = ['Freddie was a great singer', "Michael was a great singer"]
another_similar = ['Freddie was a great singer', "Freddie was a good piano player"]
typical_retrieval = ['Freddie was a great singer', "Freddie Mercury[2] (born Farrokh Bulsara; 5 September 1946 – 24 November 1991)[3] was a British singer and songwriter who achieved worldwide fame as the lead vocalist and pianist of the rock band Queen. Regarded as one of the greatest singers in the history of rock music, he was known for his flamboyant stage persona and four-octave vocal range. Mercury defied the conventions of a rock frontman with his theatrical style, influencing the artistic direction of Queen."]
long_to_long = ["Freddie Mercury[2] (born Farrokh Bulsara; 5 September 1946 – 24 November 1991)[3] was a British singer and songwriter who achieved worldwide fame as the lead vocalist and pianist of the rock band Queen. Regarded as one of the greatest singers in the history of rock music, he was known for his flamboyant stage persona and four-octave vocal range. Mercury defied the conventions of a rock frontman with his theatrical style, influencing the artistic direction of Queen.", """One of Mercury's most notable performances with Queen took place at Live Aid in 1985.[3] Queen's performance at the event has since been voted by a group of music executives as the greatest live performance in the history of rock music. The results were aired on a television program called "The World's Greatest Gigs".[57][58] Mercury's powerful, sustained note during the a cappella section came to be known as "The Note Heard Round the World"."""]


list_of_rerankers = [
    "mixedbread-ai/mxbai-rerank-large-v1",
    "BAAI/bge-reranker-v2-m3",
    "cross-encoder/ms-marco-MiniLM-L-6-v2",
    "BAAI/bge-reranker-base",
]

scores_list = []

for model_name in list_of_rerankers:
    if model_name == "cross-encoder/ms-marco-MiniLM-L-6-v2":
        model = CrossEncoder(model_name, default_activation_function = nn.Sigmoid())
    else:
        model = CrossEncoder(model_name)
    
    scores = model.predict([same, similar, another_similar, typical_retrieval, long_to_long])
    scores_list.append(
        {
            'model_name': model_name,
            'same': scores[0],
            'similar': scores[1],
            'another_similar': scores[2],
            'typical_retrieval': scores[3],
            'long_to_long': scores[4]
        }
    )
```
Returning the following results:
|    | model_name                           |   same |   similar |   another_similar |   typical_retrieval |   long_to_long |
|---:|:-------------------------------------|-------:|----------:|------------------:|--------------------:|---------------:|
|  0 | mixedbread-ai/mxbai-rerank-large-v1  | 0.8989 |    0.0708 |            0.2356 |              0.9933 |         0.9256 |
|  1 | BAAI/bge-reranker-v2-m3              | 1      |    0.0458 |            0.0589 |              0.9955 |         0.9019 |
|  2 | cross-encoder/ms-marco-MiniLM-L-6-v2 | 0.9999 |    0.5003 |            0.9846 |              0.9997 |         0.8321 |
|  3 | BAAI/bge-reranker-base               | 1      |    0.1154 |            0.5492 |              0.9729 |         0.9788 |

**Comment by tomaarsen** [comment]
Hello!

There's a few prefaces that I'd like to make:
1. Unlike with bi-encoders, cross-encoders do not necessarily produce a score of 1 for identical text. This is counterintuitive perhaps, but cross-encoders output a logit value that usually ranges between like -25 and 25 (depending hugely on the model), and then a Sigmoid turns that into a more useful score. Identical texts usually result in a high logit value, but not necessarily one that a Sigmoid turns into 1.
2. Cross-encoders are in theory stronger than bi-encoders. However, there has been more research and work on advancing bi-encoders, such that the gap between the two is smaller than you might expect.
3. Cross-encoders are primarily trained and evaluated for retrieval. I think all of the ones on your list are.

As for your results:
1. **same**: I think this can be explained by #1 from the prefaces.
2. **similar**: I think this is best explained by how these models are primarily for retrieval instead of Semantic Textual Similarity (STS). Why the original CE model "performs better" might be because it's trained for less steps than the others, so it hasn't yet "unlearned" the STS part.
3. **another_similar**: See above, though I can't explain why the original CE is really high here and the bge-reranker-base is also at 0.5.
4. **typical_retrieval**: This is clearly what the models are trained for, working as intended here.
5. **long_to_long**: This is actually a bit surprising to me, I expected this to also be too dissimilar from the "normal" query-passage.

It's possible that the "weird" **similar** and **another_similar** behaviour is because these models are not used to the "second part" of the cross-encoder input being short. It's rather common for models to not perform well on the kind of data (sizes) that it has never seen before in training. That said, this is all just my intuition. 

In short: I think this is just a mix of the training data mixed with the sometimes odd and inexplicable nature of NLP models.

- Tom Aarsen

**Comment by krumeto** [comment]
Thank you, @tomaarsen for your intuition and explanation!

My takeaway is that, while cross-encoders are trained and tested thoroughly on the typical retrieval task (short-to-long), they might not follow expectations on other tasks. One needs to assess models on the specific non-retrieval task. One also should not assume a model that is quite powerful for retrieval (e.g.  `mxbai-rerank-large-v1`) is better at other tasks as well. 

I will leave the issue open for about a week and close it, unless there is an active discussion. Thank you again!

**Comment by aamir-s18** [comment]
@krumeto thats exactly the right take away. It really depends on training data and objective to teach the model what "relevance" or "similarity" mean. Your take away is also true for embedding models btw. It really depends on what they are optimized for and for what data. 

Cross-encoder are typically used for second stage retrieval thus a lot of folks including us optimize them exactly for that.
